pkg/operator/controller/externaldns/credentials-request.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externaldnscontroller
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	cco "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorv1alpha1 "github.com/openshift/external-dns-operator/api/v1alpha1"
+	controller "github.com/openshift/external-dns-operator/pkg/operator/controller"
+)
+
+// ensureExternalCredentialsRequest ensures that the externalDNS credential request exists.
+// Returns a boolean if the credential request exists, its current state if it exists
+// and an error if it cannot be created or updated.
+func (r *reconciler) ensureExternalCredentialsRequest(ctx context.Context, externalDNS *operatorv1alpha1.ExternalDNS) (bool, *cco.CredentialsRequest, error) {
+	name := controller.ExternalDNSCredentialsRequestName(externalDNS)
+
+	desired, err := desiredCredentialsRequest(name, externalDNS)
+	if err != nil {
+		return false, nil, err
+	}
+
+	exists, current, err := r.currentExternalDNSCredentialsRequest(ctx, name)
+
+	if err != nil {
+		return false, nil, err
+	}
+
+	if !exists {
+		if err := r.createExternalDNSCredentialsRequest(ctx, desired); err != nil {
```

Comment:
I am creating the Credentials request only if platform is OCP and if Provider Types are AWS, GCP and Azure.

Reply:
if someone specifies credentials in the CR when platform is OCP and if Provider Types are AWS, GCP and Azure then we get fetch the secret we got from CCO and those creds will be ignored.

Reply:
Continuing Arjun's question about the changes.

I think that the check for the need of the update is missing in this `ensure` function, similar to what has been done for other resources: [cluster role example](https://github.com/openshift/external-dns-operator/blob/main/pkg/operator/controller/externaldns/rbac.go#L58).
When the update function is there, `desired` variable would make sense to remain on its current place.

Reply:
Update is not needed as we will creating secrets by making a call to CCO once the CR is created. Operator watches the src ns and adds that secret  to target ns if it is missing.

Reply:
What would happen if I as cluster admin change `externaldns-credentials-request-aws` credentials request's provider spec to Azure one?

Reply:
Azure credentials request will get created. I just removed the checks as @arjunrn mentioned to do so here  https://github.com/openshift/external-dns-operator/pull/50#discussion_r746615496

Reply:
i think I shall revert that because we need checks, so if someone mistakenly says  create cred request Azure on AWS, the provider type will be checked and AWS credentials request will be created.

Reply:
I think you are mixing up the provider type of ExternalDNS custome resource and the provider spec of `Credentials Request`. What I meant was the unsolicited changes in the provider spec of `CredentialsRequest`. For the moment your code only checks the presence of `CredentialsRequest` resource by name (`Get` API call), not its content. So if the name of the resource doesn't change but only its content (e.g. `ProviderSpec`) the operator won't notice that and the source (and then the destination) secret will contain undesired (modified by the third party) content (e.g.`ProviderSpec`).

Reply:
hmm OK i will add update function.

Reply:
done